pkg/controller/master/node/node_down_controller.go

```diff
@@ -156,3 +174,89 @@ func getNodeCondition(conditions []corev1.NodeCondition, conditionType corev1.No
 	}
 	return cond
 }
+
+// deleteVolumeAttachmentsOfPod deletes all volume attachments of a pod,
+// when the pod is on a down node.
+// due to controller-manager handle too long time, cause VM restart wait a long time.
+// then we delete VolumeAttachments directly.
+// ref: https://github.com/harvester/harvester/issues/4049
+func (h *nodeDownHandler) deleteVolumeAttachmentsOfPod(pod *corev1.Pod) error {
+	volumeAttachments, err := h.getVolumeAttachmentsOfPod(pod)
+	if err != nil {
+		return err
+	}
+	for _, va := range volumeAttachments {
+		if va.DeletionTimestamp == nil {
+			err := h.vas.Delete(va.Name, &metav1.DeleteOptions{})
+			if err != nil {
+				if errors.IsNotFound(err) {
+					continue
+				}
+				return err
+			}
+			logrus.Infof("deleted volume attachment %v for pod %v on downed node %v",
+				va.Name,
+				pod.Name,
+				pod.Spec.NodeName)
+		}
+	}
+	return nil
+}
+
+func (h *nodeDownHandler) getVolumeAttachmentsOfPod(pod *corev1.Pod) ([]*storagev1.VolumeAttachment, error) {
+	res := make([]*storagev1.VolumeAttachment, 0)
+	volumeAttachments, err := h.vaCache.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	pvs := make(map[string]bool)
+
+	for _, vol := range pod.Spec.Volumes {
+		if vol.VolumeSource.PersistentVolumeClaim == nil {
```

Comment:
@WebberHuang1118 's suggestion is practical,

The code could be optimized as:

(1) In the caller (NODE level), list `volumeAttachments` first, say `vas`, better to check if could list by nodeName;  after that, when `len==0`, no need to call this function,

(2) In this function, generate `pvs` first (most PODs do not have PVC), when  `len != 0`, then compare `pvs` and `vas`

You can also exchange the sequence of (1) and (2), get a POD list which do have PVC first, then get a list of `vas` is this NODE, finally cross check them.

Reply:
In this function, filter out VM pods that should have one or more PVCs.